File: 79_Word Search.java
class Solution {
    private int row, col;
    private int[][] dir = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

    public boolean exist(char[][] board, String word) {
        row = board.length;
        col = board[0].length;

        for (int r = 0; r < row; r++) {
            for (int c = 0; c < col; c++) {
                if (dfs(board, word, r, c, 0)) return true;
            }
        }

        return false;
    }

    private boolean dfs(char[][] board, String word, int curRow, int curCol, int wIndex) {
        if (wIndex == word.length()) return true;

        if (curRow < 0 || curRow >= row || curCol < 0 || curCol >= col || word.charAt(wIndex) != board[curRow][curCol]) return false;

        board[curRow][curCol] = '#';
        boolean res = false;
        for (int[] d : dir) {
            int tmpRow = curRow + d[0], tmpCol = curCol + d[1];
            res = res || (dfs(board, word, tmpRow, tmpCol, wIndex + 1));
        }

        board[curRow][curCol] = word.charAt(wIndex);
        return res;
    }
}

File: 79_Word Search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        dirs = [(-1, 0), (0, -1), (0, 1), (1, 0)]
        row, col = len(board), len(board[0])

        def dfs(curR, curC, idx) -> bool:
            # end case
            if curR < 0 or curR >= row or curC < 0 or curC >= col or board[curR][curC] != word[idx]: return False
            if idx == len(word) - 1: return True

            # process
            bt_mark = board[curR][curC]
            board[curR][curC] = '.'

            for dr, dc in dirs:
                if dfs(curR + dr, curC + dc, idx + 1): return True
            
            board[curR][curC] = bt_mark
            return False
        
        for r in range(row):
            for c in range(col):
                if dfs(r, c, 0): return True
        return False